pypose/module/ipddp.py

```diff
@@ -0,0 +1,573 @@
+        >>> S = torch.tile(torch.zeros(ns, nc, device=device), (n_batch, T, 1, 1))
+        >>> c = torch.tile(torch.zeros(1, device=device), (n_batch, T))
+        >>> stage_cost = pp.module.QuadCost(Q, R, S, c)
+        >>> terminal_cost = pp.module.QuadCost(10./dt*Q[...,0:1,:,:], R[...,0:1,:,:], S[...,0:1,:,:], c[...,0:1]) # special stagecost with T=1
+        >>>
+        >>> gx = torch.tile(torch.zeros( 2*nc, ns, device=device), (n_batch, T, 1, 1))
+        >>> gu = torch.tile(torch.vstack( (torch.eye(nc, nc, device=device), - torch.eye(nc, nc, device=device)) ), (n_batch, T, 1, 1))
+        >>> g = torch.tile(torch.hstack( (-0.25 * torch.ones(nc, device=device), -0.25 * torch.ones(nc, device=device)) ), (n_batch, T, 1))
+        >>> lincon = pp.module.LinCon(gx, gu, g)
+        >>>
+        >>> traj_opt = [None for batch_id in range(n_batch)]
+        >>> for batch_id in range(n_batch): # use for loop and keep the IPDDP
+        >>>     stage_cost = pp.module.QuadCost(Q[batch_id:batch_id+1], R[batch_id:batch_id+1], S[batch_id:batch_id+1], c[batch_id:batch_id+1])
+        >>>     terminal_cost = pp.module.QuadCost(10./dt*Q[batch_id:batch_id+1,0:1,:,:], R[batch_id:batch_id+1,0:1,:,:], S[batch_id:batch_id+1,0:1,:,:], c[batch_id:batch_id+1,0:1])
+        >>>     lincon = pp.module.LinCon(gx[batch_id:batch_id+1], gu[batch_id:batch_id+1], g[batch_id:batch_id+1])
+        >>>     init_traj_sample = {'state': init_traj['state'][batch_id:batch_id+1], 'input': init_traj['input'][batch_id:batch_id+1]}
+        >>>     solver = IPDDP(sys, stage_cost, terminal_cost, lincon, gx.shape[-2], init_traj_sample)
+        >>>     traj_opt[batch_id] = solver.optimizer()
+
+    '''
+    def __init__(self, sys=None, stage_cost=None, terminal_cost=None, cons=None, n_cons=0, init_traj=None):
+        super().__init__()
+        self.f_fn = sys
+        self.p_fn = terminal_cost
+        self.q_fn = stage_cost
+        self.c_fn = cons
+
+        self.constraint_flag = True
+        self.contraction_flag = True
+
+        self.x, self.u = init_traj['state'], init_traj['input']
+        B = self.x.shape[:-2]
+        ns, nc, ncons, self.T = self.x.size(-1), self.u.size(-1), n_cons, self.u.size(-2)
+
+        # algorithm parameter
+        self.mu, self.maxiter, self.tol, self.infeas = 1.0, 50, torch.tensor([1.0e-7], dtype=self.x.dtype, device=self.x.device), False
+
+        # quantities in forward pass
+        self.c = torch.zeros(B + (self.T, ncons), dtype=self.x.dtype, device=self.x.device)
+        self.s = 0.1 * torch.ones(B + (self.T, ncons), dtype=self.x.dtype, device=self.x.device) # dual variable for inequality constraint c
+        self.y = 0.01 * torch.ones(B + (self.T, ncons), dtype=self.x.dtype, device=self.x.device) # slack variable for infeasible case
+
+        # -------- derivatives --------------------------
+        # terms related with system dynamics
+        self.fx = torch.zeros(B + (self.T, ns, ns), dtype=self.x.dtype, device=self.x.device)
```

Comment:
Line 208 to 237 are a bit overwhelm. I think there should be a better way without defining all of the variables at the beginning. Maybe @wang-chen can have some suggestion.